src/gfx/convert.cpp

```diff
@@ -0,0 +1,831 @@
+			int nbTransparentEntries;
+			if (png_get_tRNS(png, info, &transparencyPal, &nbTransparentEntries, nullptr)) {
+				assert(nbTransparentEntries == nbColors);
+			}
+
+			options.verbosePrint("Embedded palette has %d colors: [", nbColors);
+			for (int i = 0; i < nbColors; ++i) {
+				auto const &color = embeddedPal[i];
+				options.verbosePrint("#%02x%02x%02x%02x%s", color.red, color.green, color.blue,
+				                     transparencyPal ? transparencyPal[i] : 0xFF,
+				                     i != nbColors - 1 ? ", " : "]\n");
+			}
+		} else {
+			options.verbosePrint("No embedded palette\n");
+		}
+
+		// Set up transformations; to turn everything into RGBA888
+		// TODO: it's not necessary to uniformize the pixel data (in theory), and not doing
+		// so *might* improve performance, and should reduce memory usage.
+
+		// Convert grayscale to RGB
+		switch (colorType & ~PNG_COLOR_MASK_ALPHA) {
+		case PNG_COLOR_TYPE_GRAY:
+			png_set_gray_to_rgb(png); // This also converts tRNS to alpha
+			break;
+		case PNG_COLOR_TYPE_PALETTE:
+			png_set_palette_to_rgb(png);
+			break;
+		}
+
+		// If we read a tRNS chunk, convert it to alpha
+		if (png_get_valid(png, info, PNG_INFO_tRNS))
+			png_set_tRNS_to_alpha(png);
+		// Otherwise, if we lack an alpha channel, default to full opacity
+		else if (!(colorType & PNG_COLOR_MASK_ALPHA))
+			png_set_add_alpha(png, 0xFFFF, PNG_FILLER_AFTER);
+
+		// Scale 16bpp back to 8 (we don't need all of that precision anyway)
+		if (bitDepth == 16)
+			png_set_scale_16(png);
+		else if (bitDepth < 8)
+			png_set_packing(png);
+
+		// Set interlace handling (MUST be done before `png_read_update_info`)
+		int nbPasses = png_set_interlace_handling(png);
+
+		// Update `info` with the transformations
+		png_read_update_info(png, info);
+		// These shouldn't have changed
+		assert(png_get_image_width(png, info) == width);
+		assert(png_get_image_height(png, info) == height);
+		// These should have changed, however
+		assert(png_get_color_type(png, info) == PNG_COLOR_TYPE_RGBA);
+		assert(png_get_bit_depth(png, info) == 8);
+
+		// Now that metadata has been read, we can process the image data
+
+		std::vector<png_byte> row(png_get_rowbytes(png, info));
+
+		if (interlaceType == PNG_INTERLACE_NONE) {
+			for (png_uint_32 y = 0; y < height; ++y) {
+				png_read_row(png, row.data(), nullptr);
+
+				for (png_uint_32 x = 0; x < width; ++x) {
+					Rgba rgba(row[x * 4], row[x * 4 + 1], row[x * 4 + 2], row[x * 4 + 3]);
+
+					colors.registerColor(rgba);
+					pixel(x, y) = rgba.cgbColor();
+					isGrayOnly &= rgba.isGray();
+				}
+			}
+		} else {
+			// For interlace to work properly, we must read the image `nbPasses` times
+			for (int pass = 0; pass < nbPasses; ++pass) {
+				// The interlacing pass must be skipped if its width or height is reported as zero
+				if (PNG_PASS_COLS(width, pass) == 0 || PNG_PASS_ROWS(height, pass) == 0) {
+					continue;
+				}
```

Comment:
I've noticed some other single-statement `if`s without braces. Are we consistently doing one or the other? Can clang-format enforce it either way?

Reply:
Imo we should always use braces, but shaking off old habits is hard. I thought `clang-format` would take care of it :S